File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;

class Solution {
    public int solution(String[][] book_time) {
        
        Arrays.sort(book_time, new Comparator<String[]>(){
            @Override
            public int compare(String o1[], String o2[]){
                if(!o1[0].equals(o2[0])){
                    return o1[0].compareTo(o2[0]);
                }
                return o1[1].compareTo(o2[1]);
            }
        });
        
        PriorityQueue<Integer> room = new PriorityQueue<>();
        
        for(int i=0; i<book_time.length; i++){
            String guest[] = book_time[i];
            String start[] = guest[0].split(":");
            String end[] = guest[1].split(":");
            int stTime = Integer.parseInt(start[0])*60 + Integer.parseInt(start[1]);
            int endTime = Integer.parseInt(end[0])*60 + Integer.parseInt(end[1]);
            if(room.isEmpty()){
                room.add(endTime+10);
            }else{
                if(room.peek() <= stTime){
                    room.poll(); 
                }
                room.add(endTime+10);
            }
        }
        
        return room.size();
    }
}

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.js
function solution(book_time) {
    
    const time = Array.from({length : book_time.length}, () => Array(2).fill(0));
    
    for(let i=0; i<book_time.length; i++){
        let startTime = book_time[i][0].split(":");
        let endTime = book_time[i][1].split(":");
        time[i][0] = Number.parseInt(startTime[0])*60 + Number.parseInt(startTime[1])
        time[i][1] = Number.parseInt(endTime[0])*60 + Number.parseInt(endTime[1]) +10
    }
    
    time.sort((a, b) => {
        if(a[0] == b[0]){
            return a[1] - b[1]
        }else{
            return a[0] - b[0]
        }
    })
    
    const hotel = [];
    
    for(let i=0; i<time.length; i++){
        if(hotel.length == 0){
            hotel.push(time[i][1])
        }else{
            hotel.sort((a,b) => a-b);
            if(hotel[0] <= time[i][0]){
                hotel.shift()
            }
            hotel.push(time[i][1])
        }
    }
    
    return hotel.length;
}